core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -141,41 +141,43 @@ private void removeRecovery(ShardRouting routing) {
         addRecovery(routing, false, null);
     }
 
-    private void addInitialRecovery(ShardRouting routing, RoutingTable routingTable) {
-        addRecovery(routing, true, routingTable);
+    private void addInitialRecovery(ShardRouting routing, ShardRouting initialPrimaryShard) {
+        addRecovery(routing, true, initialPrimaryShard);
     }
 
-    private void addRecovery(final ShardRouting routing, final boolean increment, final RoutingTable routingTable) {
+    private void addRecovery(final ShardRouting routing, final boolean increment, final ShardRouting initialPrimaryShard) {
```

Comment:
nit: I initially go confused by the method name `addRecovery` as it used to update the recovery counters for both addition and removal. maybe call it updateRecoveryCounts ?